org/postgresql/jdbc2/TypeInfoCache.java

```diff
@@ -150,9 +153,14 @@ public synchronized void addCoreType(String pgTypeName, Integer oid, Integer sql
         Character delim = ',';
         _arrayOidToDelimiter.put(oid, delim);
 
-        String pgArrayTypeName = "_" + pgTypeName;
```

Comment:
Ok, I stand corrected that information_schema would help.

I thought the issue is that first select returns _int4 and the second one int4[] (due to caching).

If that test must be expecting same returned type then either droping the cached value or caching all array type aliases would achieve the same result.
Still, I'm not convinced that that test must return the same type, although I agree from point of correctness that dropping the cache seems fine (not sure what are the perf implications of that).